Document interbit-ui-components

karikushi opened the pull request:
- add a README file covering how to get started with `interbit-ui-components`
- add a CONTRIBUTING file covering how to create and modify components

Commit:
(interbit-ui-components): rename README.md to README.adoc

Commit:
(interbit-ui-components): add getting started steps for using UI components

Commit:
(interbit-ui-components): add empty doc file for how to create Interbit UI Components

Commit:
(interbit-ui-components): rename create.md to CONTRIBUTING.md

Commit:
(interbit-ui-components): add creating new component docs to contribution guide

Commit:
(interbit-ui-components): update contribution guide

Commit:
(interbit-ui-components): fix line lengths

Commit:
(interbit-ui-components): fix spelling errors

Commit:
(interbit-ui-components): fix spelling

brentbenton commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+
+== Submitting your work
+
+Create a branch on the Interbit repo and create a
```
External developers cannot create a branch. They have to fork the repository and create a pull request from their fork/branch to our repo/master.

karikushi replied:
👍

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
````diff
@@ -0,0 +1,205 @@
+- Give a description of each prop that the component accepts
+
+For example, the `LinkBar` component has the following code comments.
+```js
````
Please move any source code that needs to be included into separate files. Source code isn't spell-checked, so a number of the dictionary entries you had to make would be unnecessary.

To include the source code, use this markup:
```
[source,js]
----
{% include "sourcefile.js" %}
----
```

djlmills commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+
+| `src/components/BlockExplorer`
+| An interactive component displaying a chain's state, its blocks, and actions
+dispatched.
```
This doesn't read quite right. Replace 'displaying' with 'for displaying' or 'that displays'.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+This document is for developers who wish to contribute to the
+`interbit-ui-components` package within the `interbit` repo. For general
+information on contributing to
+link:https://github.com/interbit/interbit[Interbit], please see this
```
Don't say `please`. It's simpler to just say: For this, see that.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+| Directory or file
+| Description
+
+
```
Nit: remove extra blank line.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+
+[cols="1a,1a", options="header"]
+|===
+| Directory or file
```
Maybe just 'Path'?

djlmills commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+
+Simple, presentational components should go in the `src/components/UIKit`
+directory. More complex components should be in their own directory within
+`src/components`. Each component must be exported from `src/index.js`
```
and have a test case in `exports.test.js` to verify that it is exported and usable externally.

karikushi replied:
This is mentioned in the Tests subsection. I'll add bit more description there about the export test.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+| Description
+
+
+| `src/assets`
```
Especially if you change the heading for this column to `Path`, add a trailing slash to all directories, to help distinguish folders from files.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+| `src/components`
+| React components exported from this package.
+
+| `src/components/BlockExplorer`
```
Are we describing the file structure of an existing component, or the file layout of a to-be-created component? If the latter, are we expecting users to include the BlockExplorer in the components they create?

karikushi replied:
This table describes the existing file structure.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+
+
+== Creating components
+
```
There should be an introductory paragraph here that identifies the concerns for creating components.

djlmills commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+
+=== File structure
+
+[cols="1a,1a", options="header"]
```
When clicking 'View' in Github, this table does not display correctly. Lines are truncated and everything from 'Code Structure' onwards is in the right hand column in the table.

btlewan replied:
I've verified that the table is rendering properly, both on GitHub and in the rendered GitBook HTML.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+=== Code structure
+
+We are in the process of defining the code structure for
+Interbit UI Components. At this time, we ask that contributing developers use
```
Does `Components` need to be capitalized here?

djlmills commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
`````diff
@@ -0,0 +1,205 @@
+
+The following is `LinkBar`'s example `md` file:
+
+````
`````
This doesn't render correctly in GitHub. You see a pair of backticks.

btlewan replied:
Backticks are the "compatibility" markup for source code blocks. Ideally, we'd use Asciidoctor block demarcation:

```
[source,js]
----
Content containing backticks, source code, etc.
----
```

See: https://asciidoctor.org/docs/user-manual/#delimited-blocks

djlmills commented on reference/interbit-ui-components/README.adoc:
```diff
@@ -0,0 +1,85 @@
+
+A library of React UI components specifically designed for use with Interbit.
+
+Interbit UI Components is intended to help developers with the frontend UI
```
'frontend' is two words.

karikushi replied:
Commonly used variants are front end, front-end, and frontend (and similar for backend). Does @btlewan have a preference for which is used in our docs?
https://english.stackexchange.com/questions/34447/is-it-front-end-frontend-or-front-end

djlmills replied:
My preference is front-end

btlewan replied:
It depends on the use. If you use it as a compound adjective, as in "front-end developer", it should be hyphenated. If used as a noun (kind of), as in "Go to the front end", it's really "front" as an adjective of "end", and so: front end.

Despite the popularity of removing "unnecessary" hyphens and spaces, using such compressed words is a detriment to non-English speakers and makes any translation efforts more difficult. For example, as "frontend", it's less clear where the syllable stresses should be... is it "front" and "end", "fron" and "tend", or "fronten" and "d"?

That said, there are plenty of examples using "frontend", so I didn't question it.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+directory. More complex components should be in their own directory within
+`src/components`. Each component must be exported from `src/index.js`
+
+==== SCSS
```
I think this section should be a peer of `File structure` and `Code structure`.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+- The CSS must be supported by the latest versions of Chrome, Firefox, Safari,
+and Edge. If in doubt, see link:https://caniuse.com/[caniuse.com].
+
+==== Tests
```
I think this section should be a peer of `File structure` and `Code structure`.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+`src/tests/uiKit.test.js`. Interbit UI Components uses Jest and Enzyme for
+testing and rendering components.
+
+==== Documentation
```
I think this section should be a peer of `File structure` and `Code structure`.

djlmills commented on reference/interbit-ui-components/README.adoc:
```diff
@@ -0,0 +1,85 @@
+import { Provider } from 'react-redux'
+...
+
+import 'interbit-ui-components/dist/css/interbit.css'
```
This seems like a long-winded import. Is there any way of making it shorter (without the dist)?

karikushi replied:
At this time, no. The scss is compiled into a css file in the `dist` folder. We're thinking of hosting the css on a CDN at some point.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+     */
+    clickHandler: PropTypes.func
+  }
+...
```
To avoid copy+pasta issues, make sure that ellipsis are implemented as comments.

karikushi replied:
🍝

djlmills replied:
spaghetti code?

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
````diff
@@ -0,0 +1,205 @@
+}
+```
+
+Each UI Kit component should be accompanied by an example file,
````
We'll have to complete our implementation to determine whether this paragraph is correct or not. :smile:

djlmills replied:
should is a statement of intent. If we said must...

btlewan replied:
:+1: for David's comment.

djlmills commented on reference/interbit-ui-components/README.adoc:
```diff
@@ -0,0 +1,85 @@
+== Getting Started
+
+There are two ways to get started with Interbit UI Components. The easiest is
+to use the Interbit Template app as a starting point. You can also add Interbit
```
Using `interbit-template` is not really a starting point for Interbit UI Components. It is a starting point for building an Interbit application that happens to use Interbit UI Components. Maybe something like:

'The easiest is to use the Interbit Template app as a starting point for a new Interbit application and use Interbit UI Components to build the applications UI.'

karikushi replied:
👌

djlmills reviewed (changes requested):
Some weirdness with rendering. It could just be that Github doesn't play nice with adoc format, but users will view the docs in context in GitHub.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+
+Each UI Kit component should be accompanied by an example file,
+`MyComponent.md`. link:https://react-styleguidist.js.org/docs/documenting.html[React Styleguidist]
+will render the components and JSDoc comments in a style guide.
```
Avoid future tense: ~~will~~ renders the components...

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,205 @@
+
+The example file should include:
+- An example of the rendered component.
+- An example of each variant of the component. For example, we might apply
```
Saying `For example` in an example of an example is too many examples. Perhaps `For instance` instead?

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
````diff
@@ -0,0 +1,205 @@
+Full-sized examples. To see how the component behaves responsively, change your
+viewport width.
+
+```jsx
````
Extract to separate file, and replace with include directive (see earlier comment).

btlewan commented on reference/interbit-ui-components/README.adoc:
```diff
@@ -0,0 +1,85 @@
+
+A library of React UI components specifically designed for use with Interbit.
+
+Interbit UI Components is intended to help developers with the frontend UI
```
... ~~is~~are intended...

And again, is `Components` supposed to be capitalized?

btlewan replied:
A different option:

```
The Interbit UI Components package is intended...
```

btlewan commented on reference/interbit-ui-components/README.adoc:
```diff
@@ -0,0 +1,85 @@
+=== Starting with the Interbit Template app
+
+Interbit comes with a link:/examples/template.adoc[Template app], which can be
+used a starting point for building your own Interbit applications. Follow the
```
Follow the link:/getting-started/create.md[Getting Started] steps to create...

btlewan commented on reference/interbit-ui-components/README.adoc:
```diff
@@ -0,0 +1,85 @@
+
+=== Using Interbit UI Components in your own project
+
+Interbit UI Components is available as an `npm` package. You can install and
```
~~is~~are

btlewan replied:
Or use the `package` suggestion from earlier in this section.

btlewan commented on reference/interbit-ui-components/README.adoc:
```diff
@@ -0,0 +1,85 @@
+=== Using Interbit UI Components in your own project
+
+Interbit UI Components is available as an `npm` package. You can install and
+save the dependency in your project's `package.json` by running:
```
...file by running:

btlewan commented on reference/interbit-ui-components/README.adoc:
````diff
@@ -0,0 +1,85 @@
+npm install interbit-ui-components --save
+```
+
+Once `interbit-ui-components` is installed, import the CSS file in your
````
Once the `interbit-ui-components` package is installed, ...

btlewan commented on reference/interbit-ui-components/README.adoc:
````diff
@@ -0,0 +1,85 @@
+For example, in the Template app, the import happens in `index.js` at the root
+of the application:
+
+```js
````
Extract to separate file, and replace with include directive (see earlier comment).

btlewan commented on reference/interbit-ui-components/README.adoc:
````diff
@@ -0,0 +1,85 @@
+
+For example, in the Template app, we use the `Header` component which displays
+the Interbit logo and a navigation menu.
+```js
````
Extract to separate file, and replace with include directive (see earlier comment).

btlewan commented on reference/interbit-ui-components/README.adoc:
```diff
@@ -0,0 +1,85 @@
+
+Interbit UI Components are typical React components. To use one in your
+application, simply import it and pass in the appropriate props. Documentation
+and examples for each component are provided below.
```
The last sentence doesn't appear to be true, at least not at this time.

btlewan reviewed (changes requested):
A great start! I've provided a few comments, but it would be hard to approve until we know whether `react-styleguidist` is going to be involved or not.

Commit:
(interbit-ui-components): respond to PR feedback

djlmills reviewed (commented):
The ```` is still rendering as a pair of backticks. Pulling the code out into a separate file  may help.

Commit:
(interbit-ui-components): fix nested backticks

nicholeous commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,179 @@
+directory. More complex components should be in their own directory within
+`src/components`. Each component must be exported from `src/index.js`.
+
+=== SCSS
```
Do we want to include guidelines on images? Size restrictions, format, quality, use in code?

karikushi replied:
Yes, that's something we should specify. We're not decided on what that should be in the long run, but I'll add some guidelines that will work for now.

Commit:
(interbit-ui-components): respond to PR feedback

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+= Contributing
```
Please rename this file to `contributing.adoc`. We only use ALLCAPS filenames for GitBook-specific files.

karikushi replied:
What qualifies as a GitBook-specific file?

btlewan replied:
`SUMMARY.md`, `README.md` or `README.adoc` (following GitHub's convention if using these as index pages for each folder), and `GLOSSARY.md`. By naming the file `CONTRIBUTING.md`, the implication is that it contains instructions for contributing to the folder that contains it, whereas we're talking about contributions in an entirely separate repo.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+| Rendering tests for UI Kit components.
+
+| `src/styleguide/`
+| Build files for React Styleguidist.
```
Link to the `react-styleguidist` web site.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+| Export statements for components.
+
+| `styleguide.config.js`
+| Configuration file for React Styleguidist.
```
Link to the `react-styleguidist` web site.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+
+=== Documentation
+
+Update the JSDoc code comments and the React Styleguidist examples to reflect
```
Link `JSDoc` and `React Styleguidist` to their respective web sites.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+----
+
+Each UI Kit component should be accompanied by an example file,
+`MyComponent.md`. link:https://react-styleguidist.js.org/docs/documenting.html[React Styleguidist]
```
The `React Styleguidist` sentence is awkward. While the link label identifies the package, the link targets a specific aspect of the package. I'd suggest changing the link to the main `react-styleguidist` site, and then add another sentence:

```
See React Styleguidist's link:https://react-styleguidist.js.org/docs/documenting.html[Documenting]
guidelines for details.
```

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+
+=== Image assets
+
+Image assets such as icons and illustrations should be placed in `src/assets` and should satisfy the following:
```
This line is too long.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+
+- Icons files are placed in `src/assets/icons`.
+- Image asset files are in `svg` format wherever possible.
+- Files should not exceed 25KB in size at this time.
```
`25 KB`. Always separate the units from the value with a space.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+
+[source,md]
+----
+Full-sized examples. To see how the component behaves responsively, change your
```
Extract the Markdown example into a separate file, probably named 'component-example.markdown'. One of the doc CI checks looks for `*.md` files in the rendered HTML folder to identify files not included in `SUMMARY.md`, so the example file can't have the `.md` extension.

karikushi replied:
Thanks, I wanted to pull it out into a separate file but it was squawking at me.

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+
+=== Documentation
+
+Interbit pulls documentation out of JSDoc style code comments. For each
```
I think the right thing to say here, now that we have some infrastructure in place, is:

```
Interbit's documentation extracts JSDoc-style code comments and component
example files from React components in a `react-styleguidist`-compatible way.
Use the React Styleguidist
link:https://react-styleguidist.js.org/docs/documenting.html[Documenting]
guidelines for any new components you create. Specifically, for each component:
```

djlmills replied:
To be really picky 'Interbit's documentation **generator** extracts' or 'Interbit's documentation is generated from'. Documentation is the end product.

btlewan reviewed (changes requested):
A few more bits to adjust, but it's getting close!

btlewan commented on reference/interbit-ui-components/CONTRIBUTING.adoc:
```diff
@@ -0,0 +1,187 @@
+
+The following is `LinkBar`'s example `md` file:
+
+[source,md]
```
Also, there is no `md` syntax for the Prism highlighter. Replace `md` with `markdown`.

btlewan reviewed (changes requested):
Found one more thing to comment on. :-)

djlmills commented:
I'm good with this when Ewan is.

Commit:
(interbit-ui-components): respond to PR feedback, rename contributing.adoc, pull markdown example out into separate file

Commit:
Merge branch 'master' into ui-components

nicholeous reviewed (approved):
:smile_cat:

Commit:
(interbit-ui-components): fix line lengths

karikushi commented:
@btlewan I addressed your feedback. However, tests are still failing because of a missing word in the dictionary (see PR in the relevant repo) and because of jsx in the linkBar.markdown example file. I'm not sure how to resolve this second error...

btlewan commented:
I'm working on an update to the doc CI checks to prevent the false-positive error currently being encountered.

btlewan commented:
I've fixed the doc CI issues.

It turns out that GitBook behaves a bit oddly with `CONTRIBUTING.md`; it makes its internal links (and the resulting HTML filename) lowercase during transformation. Having so lower-cased the file, the navigation links are now broken. So, renaming the file should correct all of the remaining issues.

Commit:
(interbit-ui-components): change contributing.adoc to lowercase filename
